go-controller/cmd/ovnkube/ovnkube.go

```diff
@@ -426,159 +426,174 @@ func runOvnKube(ctx context.Context, runMode *ovnkubeRunMode, ovnClientset *util
 		return ovnnode.CleanupClusterNode(runMode.identity)
 	}
 
-	stopChan := make(chan struct{})
-	wg := &sync.WaitGroup{}
-	defer func() {
-		close(stopChan)
-		wg.Wait()
-	}()
-
-	var masterWatchFactory *factory.WatchFactory
-
-	if runMode.ovnkubeController {
-		// create factory and start the controllers asked for
-		masterWatchFactory, err = factory.NewOVNKubeControllerWatchFactory(ovnClientset.GetOVNKubeControllerClientset())
-		if err != nil {
-			return err
-		}
-		defer masterWatchFactory.Shutdown()
+	watchFactory, err := newWatchFactory(runMode, ovnClientset)
+	if err != nil {
+		return fmt.Errorf("failed to initialize watch factory: %w", err)
 	}
 
+	// there might be dependencies accross components when starting so run them
+	// in separate threads
+	wg := &sync.WaitGroup{}
+	var cancel context.CancelFunc
+	ctx, cancel = context.WithCancel(ctx)
```

Comment:
hmm so we used the same parent context but new Done channels?
so anything cancelled within each routine will only effect all the new threads we are spawning and the Done channel at L568 is actually defined outside this function by the caller of the runOvnKube ?